persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -306,101 +309,122 @@ static class PossibleExpirationNotification {
-               ps.setLong(1, timeService.wallClockTime());
-               rs = ps.executeQuery();
-               int batchSize = configuration.maxBatchSize();
-               List<PossibleExpirationNotification> list;
-
-               if (key2StringMapper instanceof TwoWayKey2StringMapper) {
-                  list = new ArrayList<>(batchSize);
-               } else {
-                  list = null;
-                  PERSISTENCE.twoWayKey2StringMapperIsMissing(TwoWayKey2StringMapper.class.getSimpleName());
-               }
-               long purgedAmount = 0;
-
-               try (PreparedStatement batchDelete = conn.prepareStatement(tableManager.getDeleteRowWithExpirationSql())) {
-                  long possibleAmount = 0;
-                  while (rs.next()) {
-                     String keyStr = rs.getString(2);
-                     batchDelete.setString(1, keyStr);
-                     long expiryTime = rs.getLong(3);
-                     batchDelete.setLong(2, expiryTime);
-                     batchDelete.addBatch();
-
-                     if (list != null) {
-                        InputStream inputStream = rs.getBinaryStream(1);
-                        MarshalledValue value = unmarshall(inputStream, marshaller);
-                        list.add(new PossibleExpirationNotification(keyStr, value));
-                     }
+         this.blockingManager.runBlocking(() -> purgeExpiredInBlockingThread(unicastProcessor), "jdbcstringstore-purge");
+         return unicastProcessor;
+      });
+   }
 
-                     if (++possibleAmount == batchSize) {
-                        purgedAmount += runBatchAndNotify(list, batchDelete, unicastProcessor);
-                        possibleAmount = 0;
+   private void purgeExpiredInBlockingThread(UnicastProcessor<MarshallableEntry<K, V>> unicastProcessor) {
+      try {
+         boolean key2StringMapperIsAvailable = key2StringMapper instanceof TwoWayKey2StringMapper;
+         if (!key2StringMapperIsAvailable) {
+            PERSISTENCE.twoWayKey2StringMapperIsMissing(TwoWayKey2StringMapper.class.getSimpleName());
+         }
+         long purgedAmount = 0L;
+         List<PossibleExpirationNotification> entriesToBeDeleted;
+         int[] deleteResults;  // after call of deleteEntries the same length as entriesToBeDeleted
+         do {
+            entriesToBeDeleted = getEntriesToBeDeleted();
+
+            if (!entriesToBeDeleted.isEmpty()) {
+               deleteResults = deleteEntries(entriesToBeDeleted);
+               if (deleteResults != null) {
+                  if (key2StringMapperIsAvailable) {
+                     if (deleteResults.length == entriesToBeDeleted.size()) {
+                        purgedAmount += doNotify(entriesToBeDeleted, deleteResults, unicastProcessor);
+                     } else {
+                        log.errorf("Expected %d entries to be deleted, but deleteresult contains only %d entries");
```

Comment:
This is missing the arguments, should be:
```suggestion
                        log.errorf("Expected %d entries to be deleted, but deleteresult contains only %d entries", entriesToBeDeleted.size(), deleteResults.length);
```

Reply:
sorry, my inspect did not show this yet